File: android/app/src/main/kotlin/dev/rohanjsh/pitcher/casting/ChromecastManager.kt
package dev.rohanjsh.pitcher.casting

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.core.net.toUri
import androidx.mediarouter.media.MediaControlIntent
import androidx.mediarouter.media.MediaRouteSelector
import androidx.mediarouter.media.MediaRouter
import com.google.android.gms.cast.CastMediaControlIntent
import com.google.android.gms.cast.MediaLoadRequestData
import com.google.android.gms.cast.MediaMetadata
import com.google.android.gms.cast.MediaSeekOptions
import com.google.android.gms.cast.MediaStatus
import com.google.android.gms.cast.framework.CastContext
import com.google.android.gms.cast.framework.CastSession
import com.google.android.gms.cast.framework.SessionManagerListener
import com.google.android.gms.cast.framework.media.RemoteMediaClient
import com.google.android.gms.common.images.WebImage
import com.google.android.gms.cast.MediaInfo as CastMediaInfo

class ChromecastManager(context: Context) : CastingPlaybackProvider {

    private val mediaRouter = MediaRouter.getInstance(context)
    private val castContext = CastContext.getSharedInstance(context)
    private val sessionManager = castContext.sessionManager

    private val mediaRouteSelector = MediaRouteSelector.Builder()
        .addControlCategory(
            CastMediaControlIntent.categoryForCast(
                CastMediaControlIntent.DEFAULT_MEDIA_RECEIVER_APPLICATION_ID
            )
        )
        .addControlCategory(MediaControlIntent.CATEGORY_REMOTE_PLAYBACK)
        .build()

    private var currentMedia: MediaInfo? = null
    private var stateChangeHandler: CastingPlaybackProvider.StateChangeHandler? = null
    private var devicesChangeHandler: DevicesChangeHandler? = null

    private val positionUpdateHandler = Handler(Looper.getMainLooper())
    private var positionUpdateRunnable: Runnable? = null
    private var isCurrentlyPlaying = false

    fun interface DevicesChangeHandler {
        fun onDevicesChanged(devices: List<CastDevice>)
    }

    private val discoveryCallback = object : MediaRouter.Callback() {
        override fun onRouteAdded(router: MediaRouter, route: MediaRouter.RouteInfo) {
            log("Route added: ${route.name}")
            notifyDevicesChanged()
        }

        override fun onRouteRemoved(router: MediaRouter, route: MediaRouter.RouteInfo) {
            log("Route removed: ${route.name}")
            notifyDevicesChanged()
        }

        override fun onRouteChanged(router: MediaRouter, route: MediaRouter.RouteInfo) {
            log("Route changed: ${route.name}")
            notifyDevicesChanged()
        }
    }

    private val sessionListener = object : SessionManagerListener<CastSession> {
        override fun onSessionStarting(session: CastSession) {
            log("Session starting...")
            notifyState(connectionState = CastingConnectionState.CONNECTING)
        }

        override fun onSessionStarted(session: CastSession, sessionId: String) {
            log("Session started: $sessionId")
            session.remoteMediaClient?.registerCallback(playbackCallback)
            notifyState(connectionState = CastingConnectionState.CONNECTED, deviceFromSession = session)
        }

        override fun onSessionStartFailed(session: CastSession, error: Int) {
            log("Session start failed: error=$error")
            notifyState(playbackState = CastingPlaybackState.ERROR, error = "Failed to connect (error: $error)")
        }

        override fun onSessionEnding(session: CastSession) {
            log("Session ending...")
            session.remoteMediaClient?.unregisterCallback(playbackCallback)
        }

        override fun onSessionEnded(session: CastSession, error: Int) {
            log("Session ended: error=$error")
            currentMedia = null
            notifyState()
        }

        override fun onSessionResuming(session: CastSession, sessionId: String) {
            log("Session resuming: $sessionId")
            notifyState(connectionState = CastingConnectionState.CONNECTING)
        }

        override fun onSessionResumed(session: CastSession, wasSuspended: Boolean) {
            log("Session resumed (wasSuspended=$wasSuspended)")
            session.remoteMediaClient?.registerCallback(playbackCallback)
            notifyState(connectionState = CastingConnectionState.CONNECTED, deviceFromSession = session)
        }

        override fun onSessionResumeFailed(session: CastSession, error: Int) {
            log("Session resume failed: error=$error")
            notifyState(playbackState = CastingPlaybackState.ERROR, error = "Failed to resume session (error: $error)")
        }

        override fun onSessionSuspended(session: CastSession, reason: Int) {
            log("Session suspended: reason=$reason")
        }
    }

    private val playbackCallback = object : RemoteMediaClient.Callback() {
        override fun onStatusUpdated() {
            val client = remoteMediaClient ?: return
            val status = client.mediaStatus ?: return

            val playbackState = status.toPlaybackState()
            val positionMs = client.approximateStreamPosition.coerceAtLeast(0)
            val durationMs = client.streamDuration.coerceAtLeast(0)
            val isPlaying = playbackState == CastingPlaybackState.PLAYING

            log("Playback status: $playbackState, position=${positionMs}ms")

            notifyState(connectionState = CastingConnectionState.CONNECTED, playbackState = playbackState,
                positionMs = positionMs, durationMs = durationMs, deviceFromSession = sessionManager.currentCastSession)

            updatePositionTimerState(isPlaying)
        }

        override fun onMetadataUpdated() {}
    }

    fun setDevicesChangeHandler(handler: DevicesChangeHandler) {
        this.devicesChangeHandler = handler
    }

    fun startDiscovery() {
        log("Starting discovery...")
        mediaRouter.addCallback(
            mediaRouteSelector,
            discoveryCallback,
            MediaRouter.CALLBACK_FLAG_PERFORM_ACTIVE_SCAN
        )
        sessionManager.addSessionManagerListener(sessionListener, CastSession::class.java)
        notifyDevicesChanged()
    }

    fun stopDiscovery() {
        log("Stopping discovery...")
        mediaRouter.removeCallback(discoveryCallback)
    }

    fun getDiscoveredDevices(): List<CastDevice> = mediaRouter.routes
        .filter { it.matchesSelector(mediaRouteSelector) && !it.isDefault && it.isEnabled }
        .map { CastDevice(id = it.id, name = it.name, provider = CastingProvider.CHROMECAST, modelName = it.description) }

    fun connect(deviceId: String) {
        val route = mediaRouter.routes.find { it.id == deviceId }
            ?: return notifyState(playbackState = CastingPlaybackState.ERROR, error = "Device not found: $deviceId").also { log("Device not found: $deviceId") }
        log("Connecting to: ${route.name}")
        mediaRouter.selectRoute(route)
    }

    override fun setStateChangeHandler(handler: CastingPlaybackProvider.StateChangeHandler) {
        this.stateChangeHandler = handler
    }

    override fun loadMedia(mediaInfo: MediaInfo, autoplay: Boolean, positionMs: Long) {
        val client = remoteMediaClient ?: return notifyState(playbackState = CastingPlaybackState.ERROR, error = "No active Chromecast session").also { log("Cannot load media: no active session") }
        log("Loading media: ${mediaInfo.title}"); currentMedia = mediaInfo
        client.load(MediaLoadRequestData.Builder().setMediaInfo(mediaInfo.toCastMediaInfo()).setAutoplay(autoplay).setCurrentTime(positionMs).build())
    }

    override fun play() { log("Play"); remoteMediaClient?.play() }
    override fun pause() { log("Pause"); remoteMediaClient?.pause() }
    override fun seek(positionMs: Long) { log("Seek to ${positionMs}ms"); remoteMediaClient?.seek(MediaSeekOptions.Builder().setPosition(positionMs).build()) }
    override fun stop() { log("Stop"); remoteMediaClient?.stop(); currentMedia = null }
    override fun setVolume(volume: Double) { log("Set volume: $volume"); remoteMediaClient?.setStreamVolume(volume) }
    fun setMuted(muted: Boolean) { log("Set muted: $muted"); remoteMediaClient?.setStreamMute(muted) }

    override fun disconnect() {
        if (sessionManager.currentCastSession == null) { log("No active session to disconnect"); return }
        log("Disconnecting..."); stopPositionUpdateTimer(); isCurrentlyPlaying = false
        sessionManager.endCurrentSession(true)
    }

    fun dispose() {
        log("Disposing...")
        stopPositionUpdateTimer()
        stopDiscovery()
        sessionManager.removeSessionManagerListener(sessionListener, CastSession::class.java)
    }

    private val remoteMediaClient: RemoteMediaClient?
        get() = sessionManager.currentCastSession?.remoteMediaClient

    private fun notifyState(connectionState: CastingConnectionState = CastingConnectionState.DISCONNECTED,
                             playbackState: CastingPlaybackState = CastingPlaybackState.IDLE,
                             positionMs: Long = 0, durationMs: Long = 0, error: String? = null, deviceFromSession: CastSession? = null) {
        val device = deviceFromSession?.castDevice?.let { CastDevice(id = it.deviceId, name = it.friendlyName, provider = CastingProvider.CHROMECAST, modelName = it.modelName) }
        stateChangeHandler?.onStateChanged(ProviderState(connectionState = connectionState, playbackState = playbackState, device = device, positionMs = positionMs, durationMs = durationMs, error = error))
    }

    private fun notifyDevicesChanged() = devicesChangeHandler?.onDevicesChanged(getDiscoveredDevices())

    private fun startPositionUpdateTimer() {
        if (positionUpdateRunnable != null) return
        log("Starting position update timer")
        positionUpdateRunnable = object : Runnable {
            override fun run() {
                val client = remoteMediaClient ?: return
                notifyState(connectionState = CastingConnectionState.CONNECTED, playbackState = CastingPlaybackState.PLAYING,
                    positionMs = client.approximateStreamPosition.coerceAtLeast(0),
                    durationMs = client.streamDuration.coerceAtLeast(0),
                    deviceFromSession = sessionManager.currentCastSession)
                positionUpdateHandler.postDelayed(this, POSITION_UPDATE_INTERVAL_MS)
            }
        }
        positionUpdateHandler.postDelayed(positionUpdateRunnable!!, POSITION_UPDATE_INTERVAL_MS)
    }

    private fun stopPositionUpdateTimer() { positionUpdateRunnable?.let { positionUpdateHandler.removeCallbacks(it); positionUpdateRunnable = null } }

    private fun updatePositionTimerState(isPlaying: Boolean) {
        when { isPlaying && !isCurrentlyPlaying -> { isCurrentlyPlaying = true; startPositionUpdateTimer() }
               !isPlaying && isCurrentlyPlaying -> { isCurrentlyPlaying = false; stopPositionUpdateTimer() } }
    }

    private fun log(message: String) = Log.d(TAG, message)

    companion object {
        private const val TAG = "ChromecastManager"
        private const val POSITION_UPDATE_INTERVAL_MS = 500L
    }
}

private fun MediaInfo.toCastMediaInfo(): CastMediaInfo {
    val metadata = MediaMetadata(
        if (mediaType == MediaType.VIDEO) MediaMetadata.MEDIA_TYPE_MOVIE
        else MediaMetadata.MEDIA_TYPE_MUSIC_TRACK
    ).apply {
        putString(MediaMetadata.KEY_TITLE, title)
        subtitle?.let { putString(MediaMetadata.KEY_SUBTITLE, it) }
        imageUrl?.let { addImage(WebImage(it.toUri())) }
    }

    return CastMediaInfo.Builder(contentUrl)
        .setStreamType(CastMediaInfo.STREAM_TYPE_BUFFERED)
        .setContentType(contentType ?: "video/mp4")
        .setMetadata(metadata)
        .apply { duration?.let { setStreamDuration(it) } }
        .build()
}

private fun MediaStatus.toPlaybackState(): CastingPlaybackState = when (playerState) {
    MediaStatus.PLAYER_STATE_IDLE -> when (idleReason) {
        MediaStatus.IDLE_REASON_FINISHED -> CastingPlaybackState.ENDED
        MediaStatus.IDLE_REASON_ERROR -> CastingPlaybackState.ERROR
        else -> CastingPlaybackState.IDLE
    }
    MediaStatus.PLAYER_STATE_BUFFERING -> CastingPlaybackState.LOADING
    MediaStatus.PLAYER_STATE_PLAYING -> CastingPlaybackState.PLAYING
    MediaStatus.PLAYER_STATE_PAUSED -> CastingPlaybackState.PAUSED
    else -> CastingPlaybackState.IDLE
}

File: ios/Runner/Casting/ChromecastManager.swift
import Foundation
import GoogleCast
import os.log

@MainActor
final class ChromecastManager: NSObject, CastingPlaybackProvider {

    typealias StateChangeHandler = (ProviderState) -> Void
    typealias DevicesChangeHandler = ([CastDevice]) -> Void

    private let castContext: GCKCastContext
    private let discoveryManager: GCKDiscoveryManager
    private let sessionManager: GCKSessionManager

    private var currentMedia: MediaInfo?
    private(set) var isConnected = false
    private(set) var isConnecting = false
    private var onStateChange: StateChangeHandler?
    private var onDevicesChange: DevicesChangeHandler?
    private var positionUpdateTimer: Timer?

    init(castContext: GCKCastContext) {
        self.castContext = castContext
        self.discoveryManager = castContext.discoveryManager
        self.sessionManager = castContext.sessionManager
        super.init()

        sessionManager.add(self)
        log("Initialized")
    }

    func setStateChangeHandler(_ handler: @escaping StateChangeHandler) {
        self.onStateChange = handler
    }

    func setDevicesChangeHandler(_ handler: @escaping DevicesChangeHandler) {
        self.onDevicesChange = handler
    }

    func startDiscovery() {
        discoveryManager.add(self)
        discoveryManager.passiveScan = true
        discoveryManager.startDiscovery()
        log("Discovery started")
        refreshDevices()
    }

    func stopDiscovery() {
        discoveryManager.stopDiscovery()
        discoveryManager.remove(self)
        log("Discovery stopped")
    }

    func getDiscoveredDevices() -> [CastDevice] {
        (0..<discoveryManager.deviceCount).map { index in
            let d = discoveryManager.device(at: index)
            return CastDevice(id: d.uniqueID, name: d.friendlyName ?? d.uniqueID, provider: .chromecast, modelName: d.modelName)
        }
    }

    func connect(deviceId: String) {
        guard let device = discoveryManager.device(withUniqueID: deviceId) else {
            log("Device not found: \(deviceId)")
            notifyState(error: "Device not found: \(deviceId)")
            return
        }

        log("Connecting to: \(device.friendlyName ?? deviceId)")
        isConnecting = true
        sessionManager.startSession(with: device)
        notifyState()
    }

    func disconnect() {
        guard sessionManager.hasConnectedSession() else {
            log("No active session to disconnect")
            return
        }

        log("Disconnecting...")
        stopPositionUpdateTimer()
        sessionManager.endSessionAndStopCasting(true)
    }

    func loadMedia(_ mediaInfo: MediaInfo, autoplay: Bool, positionMs: Int64) {
        guard let remoteMediaClient = remoteMediaClient else {
            log("Cannot load media: no active session")
            notifyState(error: "No active Chromecast session")
            return
        }

        guard let castMediaInfo = mediaInfo.toCastMediaInfo() else {
            log("Cannot load media: invalid content URL")
            notifyState(error: "Invalid media URL: \(mediaInfo.contentUrl)")
            return
        }

        log("Loading media: \(mediaInfo.title)")
        currentMedia = mediaInfo

        let mediaLoadOptions = GCKMediaLoadOptions()
        mediaLoadOptions.autoplay = autoplay
        mediaLoadOptions.playPosition = TimeInterval(positionMs) / 1000.0

        remoteMediaClient.loadMedia(castMediaInfo, with: mediaLoadOptions)
    }

    func play() {
        log("Play")
        remoteMediaClient?.play()
    }

    func pause() {
        log("Pause")
        remoteMediaClient?.pause()
    }

    func seek(to positionMs: Int64) {
        log("Seek to \(positionMs)ms")
        let seekOptions = GCKMediaSeekOptions()
        seekOptions.interval = TimeInterval(positionMs) / 1000.0
        remoteMediaClient?.seek(with: seekOptions)
    }

    func stop() {
        log("Stop")
        remoteMediaClient?.stop()
        currentMedia = nil
    }

    func setVolume(_ volume: Float) {
        log("Set volume: \(volume)")
        remoteMediaClient?.setStreamVolume(volume)
    }

    func setMuted(_ muted: Bool) {
        log("Set muted: \(muted)")
        remoteMediaClient?.setStreamMuted(muted)
    }

    var connectedDeviceName: String? {
        sessionManager.currentCastSession?.device.friendlyName
    }

    var connectedDeviceId: String? {
        sessionManager.currentCastSession?.device.uniqueID
    }

    private var remoteMediaClient: GCKRemoteMediaClient? {
        sessionManager.currentCastSession?.remoteMediaClient
    }

    private func refreshDevices() {
        let devices = getDiscoveredDevices()
        log("Discovered \(devices.count) Chromecast device(s)")
        onDevicesChange?(devices)
    }

    private func notifyState(error: String? = nil) {
        let connectionState: CastingConnectionState = isConnected ? .connected : isConnecting ? .connecting : .disconnected
        let playbackState: CastingPlaybackState = error != nil ? .error : isPlaying ? .playing : currentPositionMs > 0 ? .paused : .idle
        let device = connectedDeviceId.flatMap { id in connectedDeviceName.map { CastDevice(id: id, name: $0, provider: .chromecast, modelName: nil) } }
        onStateChange?(ProviderState(connectionState: connectionState, playbackState: playbackState, device: device,
                                     positionMs: currentPositionMs, durationMs: currentDurationMs, error: error))
    }

    private var isPlaying: Bool {
        guard let status = remoteMediaClient?.mediaStatus else { return false }
        return status.playerState == .playing
    }

    private var currentPositionMs: Int64 {
        guard let client = remoteMediaClient else { return 0 }
        return Int64(client.approximateStreamPosition() * 1000)
    }

    private var currentDurationMs: Int64 {
        guard let mediaInfo = remoteMediaClient?.mediaStatus?.mediaInformation else { return 0 }
        return Int64(mediaInfo.streamDuration * 1000)
    }

    private func cleanup() {
        stopPositionUpdateTimer(); sessionManager.remove(self); remoteMediaClient?.remove(self); log("Cleaned up")
    }

    private func startPositionUpdateTimer() {
        guard positionUpdateTimer == nil else { return }
        log("Starting position update timer")
        positionUpdateTimer = Timer.scheduledTimer(withTimeInterval: 0.5, repeats: true) { [weak self] _ in
            Task { @MainActor in self?.notifyState() }
        }
    }

    private func stopPositionUpdateTimer() { positionUpdateTimer?.invalidate(); positionUpdateTimer = nil }

    private func updatePositionTimerState() { isPlaying ? startPositionUpdateTimer() : stopPositionUpdateTimer() }

    private nonisolated func log(_ message: String) {
        os_log("[ChromecastManager] %{public}@", log: .default, type: .debug, message)
    }
}

extension ChromecastManager: GCKDiscoveryManagerListener {

    nonisolated func didInsert(_ device: GCKDevice, at index: UInt) {
        Task { @MainActor in
            log("Device added: \(device.friendlyName ?? device.uniqueID)")
            refreshDevices()
        }
    }

    nonisolated func didRemove(_ device: GCKDevice, at index: UInt) {
        Task { @MainActor in
            log("Device removed: \(device.friendlyName ?? device.uniqueID)")
            refreshDevices()
        }
    }

    nonisolated func didUpdate(_ device: GCKDevice, at index: UInt) {
        Task { @MainActor in
            log("Device updated: \(device.friendlyName ?? device.uniqueID)")
            refreshDevices()
        }
    }
}

extension ChromecastManager: GCKSessionManagerListener {

    nonisolated func sessionManager(_ sessionManager: GCKSessionManager, willStart session: GCKSession) {
        Task { @MainActor in
            log("Session starting...")
            isConnecting = true
            isConnected = false
            notifyState()
        }
    }

    nonisolated func sessionManager(_ sessionManager: GCKSessionManager, didStart session: GCKSession) {
        Task { @MainActor in
            log("Session started: \(session.sessionID ?? "unknown")")
            isConnecting = false
            isConnected = true

            if let castSession = session as? GCKCastSession {
                castSession.remoteMediaClient?.add(self)
            }

            notifyState()
        }
    }

    nonisolated func sessionManager(_ sessionManager: GCKSessionManager, didFailToStart session: GCKSession, withError error: Error) {
        Task { @MainActor in
            log("Session start failed: \(error.localizedDescription)")
            isConnecting = false
            isConnected = false
            notifyState(error: "Failed to connect: \(error.localizedDescription)")
        }
    }

    nonisolated func sessionManager(_ sessionManager: GCKSessionManager, willEnd session: GCKSession) {
        Task { @MainActor in
            log("Session ending...")

            if let castSession = session as? GCKCastSession {
                castSession.remoteMediaClient?.remove(self)
            }
        }
    }

    nonisolated func sessionManager(_ sessionManager: GCKSessionManager, didEnd session: GCKSession, withError error: Error?) {
        Task { @MainActor in
            if let error = error {
                log("Session ended with error: \(error.localizedDescription)")
            } else {
                log("Session ended")
            }

            isConnecting = false
            isConnected = false
            currentMedia = nil
            notifyState()
        }
    }

    nonisolated func sessionManager(_ sessionManager: GCKSessionManager, willResumeCastSession session: GCKCastSession) {
        Task { @MainActor in
            log("Session resuming...")
            isConnecting = true
            notifyState()
        }
    }

    nonisolated func sessionManager(_ sessionManager: GCKSessionManager, didResumeCastSession session: GCKCastSession) {
        Task { @MainActor in
            log("Session resumed")
            isConnecting = false
            isConnected = true
            session.remoteMediaClient?.add(self)
            notifyState()
        }
    }

    nonisolated func sessionManager(_ sessionManager: GCKSessionManager, didSuspend session: GCKSession, with reason: GCKConnectionSuspendReason) {
        Task { @MainActor in
            log("Session suspended: reason=\(reason.rawValue)")
        }
    }
}

extension ChromecastManager: GCKRemoteMediaClientListener {

    nonisolated func remoteMediaClient(_ client: GCKRemoteMediaClient, didUpdate mediaStatus: GCKMediaStatus?) {
        Task { @MainActor in
            guard mediaStatus != nil else { return }
            log("Playback status updated")
            notifyState()
            updatePositionTimerState()
        }
    }
}

extension MediaInfo {

    func toCastMediaInfo() -> GCKMediaInformation? {
        guard let url = URL(string: contentUrl) else {
            return nil
        }

        let metadata = GCKMediaMetadata(metadataType: mediaType == .video ? .movie : .musicTrack)
        metadata.setString(title, forKey: kGCKMetadataKeyTitle)

        if let subtitle = subtitle {
            metadata.setString(subtitle, forKey: kGCKMetadataKeySubtitle)
        }

        if let imageUrl = imageUrl, let imageURL = URL(string: imageUrl) {
            metadata.addImage(GCKImage(url: imageURL, width: 480, height: 270))
        }

        let builder = GCKMediaInformationBuilder(contentURL: url)
        builder.streamType = .buffered
        builder.contentType = contentType ?? "video/mp4"
        builder.metadata = metadata

        if let duration = duration {
            builder.streamDuration = TimeInterval(duration) / 1000.0
        }

        return builder.build()
    }
}

